tracing-appender/src/rolling.rs

```diff
@@ -558,17 +566,90 @@ impl Inner {
                     .unwrap_or(0),
             ),
             rotation,
+            keep_last: None,
         };
         Ok((inner, writer))
     }
 
+    fn prune_old_logs(&self, keep_last: usize) {
+        let format;
+        match self.rotation {
+            Rotation::MINUTELY => {
+                format = format_description::parse("[year]-[month]-[day]-[hour]-[minute]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::HOURLY => {
+                format = format_description::parse("[year]-[month]-[day]-[hour]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::DAILY => {
+                format = format_description::parse("[year]-[month]-[day]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::NEVER => {
+                unreachable!("keep last N files feature is not available for `Rotation::NEVER`!")
+            }
+        }
+
+        let files = fs::read_dir(&self.log_directory).map(|dir| {
+            dir.filter_map(Result::ok)
+                .filter(|entry| {
+                    entry.file_name().to_string_lossy().contains(
+                        &self
+                            .log_filename_prefix
+                            .clone()
+                            .expect("prefix is always present"),
+                    )
+                })
+                .collect::<Vec<_>>()
+        });
+
+        match files {
+            Ok(files) => {
+                if files.len() >= keep_last {
+                    let mut sorted_files = vec![];
+                    for file in files {
+                        let filename = file.file_name().to_string_lossy().to_string();
+                        match filename.rfind('.') {
+                            Some(date_index) => {
+                                let date_str = filename[(date_index + 1)..].to_string();
+                                match Date::parse(&date_str, &format) {
+                                    Ok(date) => sorted_files.push((file.path(), date)),
+                                    Err(error) => eprintln!("error parsing the date: {error}"),
+                                }
+                            }
+                            None => {
+                                eprintln!("Filename do not meet the expected format!");
+                                continue;
+                            }
+                        }
+                    }
+                    sorted_files.sort_by_key(|item| item.1);
+                    // delete files, so that (n-1) files remain, because we will create another log file
+                    for (file, _) in &sorted_files[..sorted_files.len() - (keep_last - 1)] {
```

Comment:
it looks like we're currently determining the age of a log file by reading the filename and attempting to parse the date from the filename, which feels a bit complex and potentially flaky. Instead, I think we could just use the file's creation timestamp, as returned by [`fs::Metadata::created`](https://doc.rust-lang.org/stable/std/fs/struct.Metadata.html#method.created).

Something like this could potentially be a lot simpler:
```suggestion
            Ok(mut files) => {
                if files.len() >= keep_last {
                    // sort the files by their creation timestamps.
                    files.sort_by_key(|file| {
                        file.metadata()
                            .and_then(|metadata| metadata.created().map(Some))
                            .unwrap_or_else(|error| {
                                eprintln!(
                                    "Unable to get file creation time for {}: {}",
                                    file.path().display(),
                                    error
                                );
                                // if we couldn't determine the file's creation
                                // timestamp for whatever reason, it should
                                // compare as "less than" other files, so that
                                // it's not deleted.
                                None
                            })
                    });

                    // delete files, so that (n-1) files remain, because we will create another log file
                    for file in &files[..files.len() - (keep_last - 1)] {
```

Reply:
did not know about metadata... Great suggestion thanks!